documentation/projects/proposals/infrastructure_improvements/20240105_implementation_plan_move_remaining_services_to_openverse_org.md

```diff
@@ -0,0 +1,738 @@
+each host and direct requests wherever it wants, with listeners ordered in
+arbitrary priority. That means all the API needs, at the most basic level, is to
+have new CNAME records in Cloudflare added to point `api.openverse.org` to the
+load balancer, and new load balancer listener rules to forward those requests to
+the existing API services. By itself, that would indeed make the API accessible
+on `openverse.org`. However, it would leave all the Cloudflare configuration for
+the API on `openverse.engineering` inoperable for `openverse.org` (because these
+rules are per-zone). So, the rest of the steps in this section are aimed at
+carefully preparing and moving the service's Cloudflare configuration to
+`openverse.org`.
+
+```{note}
+I am using `api(-(production|staging))?.` as shorthand for the three subdomains used
+for our staging and production API. For clarity, this should be understood to expand to:
+- `api.`
+- `api-production.`
+- `api-staging.`
+```
+
+1. Add `api(-(production|staging))?.openverse.org/admin` to Cloudflare Access
+   and duplicate all API Cloudflare rules from `openverse.engineering` to
+   `openverse.org`.
+   - [This must all be in the `cloudflare` root module](#moving-cloudflare-access-and-the-new-cloudflare-root-module).
+1. Add `api(-(production|staging))?.openverse.org` to `domain_aliases` for the
+   API services and use a new `generic/service-domain` to point
+   `api(-(production|staging))?.openverse.org` to the existing API services.
+   - Rule priority does not matter here because our rules match on the requested
+     host name, so the `openverse.org` and `openverse.engineering` listeners
+     will never interrupt each other.
+1. Redirect `api(-production)?.openverse.engineering` to `api.openverse.org`
+   using a Cloudflare
+   [dynamic single redirect rule](https://developers.cloudflare.com/rules/url-forwarding/single-redirects/)
+   that matches only when a special header is present for testing.
+   - Configure this in
+     [the `cloudflare` root module](#moving-cloudflare-access-and-the-new-cloudflare-root-module),
+     not in the environment root modules.
+   - We will redirect both `api.openverse.engineering` and
+     `api-production.openverse.engineering` to `api.openverse.org`. This reduces
+     the complexity of the dynamic target expression.
+   - We will not redirect staging. Staging has no guarantees for users.
+     Excluding it from the redirects also reduces the overall complexity of both
+     the redirect match and dynamic target expressions.
+1. Update UptimeRobot monitors to point to the `openverse.org` domains.
+1. **Production only steps**
+   1. Update the frontend to point to `api.openverse.org` (only possible after
+      configuring the domain for the production API because the staging and
+      production frontend both point to the production API).
+   1. Open PRs in Gutenberg and Jetpack to update the Openverse integration URL
+      to `api.openverse.org`
```

Comment:
Thank you for including this step explicitly! Maybe we should mention other integrations more broadly? This isn't expansive but we could also submit PRs for folks using the old URL in GitHub: https://github.com/search?type=code&q=api.openverse.engineering

Reply:
That's a nice bonus but I am hesitant to add it as a requirement to the project. At most I'd accept the idea of opening issues in those repositories, with Openverse maintainers optionally opening PRs if they feel personally motivated to do so for one reason or another.

I am hesitant to add anything to this project that we don't have control over or at least very good, clear relationahips with the other maintainers to be able to set expectations about completion.

Reply:
That's a great point, I think making issues is good but also agree it would not be a requirement.